src/WebJobs.Script.Grpc/Channel/GrpcWorkerChannel.cs

```diff
@@ -873,10 +874,34 @@ protected virtual void Dispose(bool disposing)
 
         public void Dispose()
         {
+            StopWorkerProcessAsync();
             _disposing = true;
             Dispose(true);
         }
 
+        public async void StopWorkerProcessAsync()
+        {
+            bool capabilityEnabled = !string.IsNullOrEmpty(_workerCapabilities.GetCapabilityState(RpcWorkerConstants.HandlesWorkerTerminateMessage));
+            if (!capabilityEnabled)
+            {
+                return;
+            }
+
+            var workerTerminate = new WorkerTerminate()
+            {
+                GracePeriod = Duration.FromTimeSpan(TimeSpan.FromSeconds(WorkerConstants.WorkerTerminateGracePeriodInSeconds))
+            };
+
+            _workerChannelLogger.LogDebug($"Sending WorkerTerminate message with grace period {WorkerConstants.WorkerTerminateGracePeriodInSeconds} seconds.");
+
+            SendStreamingMessage(new StreamingMessage
+            {
+                WorkerTerminate = workerTerminate
+            });
+
+            await Task.Delay(TimeSpan.FromSeconds(WorkerConstants.WorkerTerminateGracePeriodInSeconds));
```

Comment:
So we want to delay the `Dispose(true)` method call 10 seconds?, not doing the wait in the `Process.WaitForExit` code path anymore?

Reply:
Do we really need to wait for 10 seconds here? The fact that this code path is called form the Dispose method means that some of the components are already shut down/not live any more. In that case,  do we expect the host to process any messages coming from the worker during that 10 seconds timeframe?